proposals/0046-optimistic-cluster-restart-automation.md

```diff
@@ -0,0 +1,408 @@
+
+Any slot above this line should be repaired, while other slots can be ignored
+for now. But if
+`OPTIMISTIC_CONFIRMED_THRESHOLD` - `MALICIOUS_SET` - `PERCENT_NOT_IN_RESTART`
+is less than 10%, then the validators don't have to start any repairs. Because
+the signal now is too noisy.
+
+We obviously want to repair all blocks above `OPTIMISTIC_CONFIRMED_THRESHOLD`
+before the restart. The validators in `MALICIOUS_SET` could lie about their
+votes, so we need to be conservative and lower the line accordingly. Also,
+we don't know what the validators not in restart have voted, so we need to
+be even more conservative and assume they voted for this block. Being
+conservative means we might repair blocks which we didn't need, but we will
+never miss any block we should have repaired.
+
+Next we illustrate the system behavior using concrete numbers:
+
+* 5% validators in restart: `PERCENT_NOT_IN_RESTART` is 100% - 5% = 95%.
+`OPTIMISTIC_CONFIRMED_THRESHOLD` - `MALICIOUS_SET` - `PERCENT_NOT_IN_RESTART`
+= 67% - 5% - 95% < 10%, so no validators would repair any block.
+
+* 70% validators in restart: `PERCENT_NOT_IN_RESTART` is 100% - 70% = 30%.
+`OPTIMISTIC_CONFIRMED_THRESHOLD` - `MALICIOUS_SET` - `PERCENT_NOT_IN_RESTART`
+= 67% - 5% - 30% = 32%, so slots with above 32% votes accumulated from
+`RestartLastVotedForkSlots` would be repaired.
+
+* 80% validators are in restart, `PERCENT_NOT_IN_RESTART` is 100% - 80% = 20%.
+`OPTIMISTIC_CONFIRMED_THRESHOLD` - `MALICIOUS_SET` - `PERCENT_NOT_IN_RESTART`
+= 67% - 5% - 20% = 42%, so slots with above 42% votes accumulated from
+`RestartLastVotedForkSlots` would be repaired.
+
+From above examples, we can see the "must-have" threshold changes dynamically 
+depending on how many validators are in restart. The main benefit is that a
+block will only move from "must-have" to "ignored" as more validators 
+join the restart, not vice versa. So the list of blocks a validator needs to
+repair will never grow bigger when more validators join the restart.
+
+Once the validator gets `RestartLastVotedForkSlots`, it can calculate which
+blocks must be repaired. When all those "must-have" blocks are repaired and
+replayed, it can proceed to step 3.
+
+3. **gossip current heaviest fork**
+
+The main goal of this step is to "vote" the heaviest fork to restart from.
+
+We use a new gossip message `RestartHeaviestFork`, its fields are:
+
+* `slot`: `u64` slot of the picked block.
+* `hash`: `Hash` bank hash of the picked block.
+* `stake_committed_percent`: `u16` total percentage of stakes of the validators
+it received `RestartHeaviestFork` messages from.
+
+After receiving `RestartLastVotedForkSlots` from the validators holding stake
+more than `RESTART_STAKE_THRESHOLD` and repairing slots in "must-have"
+category, replay all blocks and pick the heaviest fork by traversing from
+local root like this:
+
+1. If a block has more than 67% stake in `RestartLastVotedForkSlots`
+   messages, traverse down this block.
```

Comment:
I think its important to specify here that votes for children count towards the parent as well

Reply:
Changed.